runtime/lua/vim/filetype.lua

```diff
@@ -2376,8 +2361,28 @@ local function dispatch(ft, path, bufnr, ...)
   end
 end
 
+-- Caches calls to string.find
+local expand_env_lookup = {}
+
 ---@private
 local function match_pattern(name, path, tail, pat)
+  if expand_env_lookup[pat] == nil then
+    expand_env_lookup[pat] = pat:find('%$') ~= nil
```

Comment:
This should check for `${` (or maybe even the full `%${(%S-)}`), because a plain `$` could be used to mark the end of a pattern, e.g. `vimrc$`.

```suggestion
    expand_env_lookup[pat] = pat:find('%${') ~= nil
```